src/compat/crypto.rs

```diff
@@ -4,73 +4,104 @@
-        Err(_) => return utils::fail(),
-    };
-
-    let mut is_owned: bool = false;
-    let mut nullifier_found = BlsScalar::default();
-    let mut pk_found: Option<PublicKey> = None;
+    let mut leaf_chunk = leaves.chunks_exact(TREE_LEAF_SIZE);
+    let mut last_pos = 0;
 
-    for idx in 0..=MAX_KEY {
-        let idx = idx as u64;
-        let sk = key::derive_sk(&seed, idx);
-        let vk = ViewKey::from(&sk);
+    let mut notes = Vec::new();
+    let mut nullifiers = Vec::new();
+    let mut block_heights = Vec::new();
+    let mut public_spend_keys = Vec::new();
 
-        if vk.owns(&note) {
-            let nullifier = note.gen_nullifier(&sk);
-
-            nullifier_found = nullifier;
-            is_owned = true;
-            pk_found = Some(PublicKey::from(&sk));
+    for leaf_bytes in leaf_chunk.by_ref() {
+        let TreeLeaf { block_height, note } =
+            match rkyv::from_bytes(leaf_bytes).ok() {
+                Some(a) => a,
+                None => {
+                    return utils::fail();
+                }
+            };
 
-            break;
+        last_pos = core::cmp::max(last_pos, *note.pos());
+
+        for idx in 0..=MAX_KEY {
+            let idx = idx as u64;
+            let view_key = key::derive_vk(&seed, idx);
+
+            if view_key.owns(&note) {
+                let sk = key::derive_sk(&seed, idx);
+                let nullifier = note.gen_nullifier(&sk);
+
+                let nullifier_found =
+                    match rkyv::to_bytes::<BlsScalar, MAX_LEN>(&nullifier).ok()
+                    {
+                        Some(n) => n.to_vec(),
+                        None => return utils::fail(),
+                    };
+
+                let psk_found =
+                    bs58::encode(PublicKey::from(sk).to_bytes()).into_string();
+
+                let raw_note: Vec<u8> =
+                    match rkyv::to_bytes::<Note, MAX_LEN>(&note).ok() {
+                        Some(n) => n.to_vec(),
+                        None => return utils::fail(),
+                    };
+
+                notes.push(raw_note.to_owned());
+                block_heights.push(block_height);
+                public_spend_keys.push(psk_found);
+                nullifiers.push(nullifier_found);
+            }
         }
     }
 
-    let pk_found = pk_found.map(|pk| bs58::encode(pk.to_bytes()).into_string());
-
-    let nullifier_found =
-        match rkyv::to_bytes::<BlsScalar, MAX_LEN>(&nullifier_found).ok() {
-            Some(n) => n.to_vec(),
-            None => return utils::fail(),
-        };
+    let block_heights = block_heights
+        .iter()
+        .map(|x| x.to_string())
+        .collect::<Vec<_>>()
+        .join(",");
 
     utils::into_ptr(types::CheckNoteOwnershipResponse {
-        is_owned,
-        nullifier: nullifier_found,
-        public_key: pk_found,
+        notes,
+        block_heights,
+        public_spend_keys,
+        nullifiers,
+        last_pos,
```

Comment:
We also need to redo this. Not in this PR obviously, no worries. :)

But the fact that we're still using strings (the list of block heights as string is just plain wrong, we should have them as vec of numbers or bytes), it still implies some form of additional serialization.